### ISSUE TITLE
Prometheus Apps Discovery

### ISSUE THREAD

**Issue by brust** [tsuru/tsuru#1554]
We have apps instrumentation to expose software metrics on `/metrics` endpoint. How can Prometheus get `/metrics` for each unit container? Is there a way for that?

Units containers may have different behaviors, so it's needful to get `/metrics` directly, not by PlanB.

Prometheus has Kubernetes discovery. How can we create Tsuru discovery?

**Comment by andrestc** [comment]
Currently, there is no built-in solution for this, and it's very unlikely that we will get a `tsuru discovery` merged on prometheus core. 

It should be fairly easy to build a small worker that fetches the `/apps` route on tsuru API to get every units addresses. This worker would then write a json file to be used by [prometheus file discovery](https://prometheus.io/docs/operating/configuration/#<file_sd_config>). 

Another possible solution is to write a small HTTP server that implements our service API. You could bind apps that expose the `/metrics` to these service. The service would be called everytime a unit is added/remove and would be able to keep the json file up to date.

**Comment by andrestc** [comment]
Hi, @brust, i've decided to do a little experimenting since I'm interested in this aswell. I was able to build a POC using [jq](https://stedolan.github.io/jq/) and [tsuru curl](https://gist.github.com/cezarsa/f1ad4c5bcf5ffb5418e8).

This gets all apps and output their units address (only the ones with the `web` process, adding a `/metrics` suffix) and also adds a prometheus label to each unit with its application name.
``` bash
tsuru curl /apps | jq -c '[.[] | .units | .[] | select(.ProcessName == "web") | {"targets": [.Address.Host+"/metrics"], "labels":{"appname":.AppName}}]'
```

The output is something like:
```json
[
   {
      "targets":[
         "127.0.0.1:32769/metrics",
         "127.0.0.1:32768/metrics"
      ],
      "labels":{
         "appname":"tsuru-dashboard"
      }
   },
   {
      "targets":[
         "127.0.0.1:32770/metrics"
      ],
      "labels":{
         "appname":"static"
      }
   }
]
```
This has the disadvantage that every application will be targets for scrapping (even those without the `/metrics` endpoint). With the addition of application tags(#1550, #1556), to be released on tsuru 1.3, you will be able to filter applications based on tags and could use a tag to point what applications should be scrapped by prometheus, easily filtering them on the `/apps` endpoint.

You can run something like this every minute and write its output to a directory where prometheus fetches the config (he does the diffing between the current one and the new one).

**Comment by brust** [comment]
@andrestc Thanks, I'll try it!
